Removed ProtocolLib + Resource pack load during Paper's configuration

FaultyFunctions opened the pull request:
It was a bit of a struggle but I finally have it working. Please let me know if there is something wrong with my code. A lot of these events and methods I've never used before so something might be off/incorrect.

There are some points of discussion I'd like to bring up and ask about:
### Permissions
I asked this is Discord but might as well put this in here in case anybody else wants to discuss. The way Paper's configuration phase works is you don't have access to the full player object but rather a [PlayerProfile](https://jd.papermc.io/paper/1.21.8/com/destroystokyo/paper/profile/PlayerProfile.html). This means it's a bit trickier to gather what permissions the player has and if they should be able to use `vane.core.resource_pack.bypass` to bypass resource pack requirements.

An alternative might be to interface directly with popular permission plugins like LuckPerms and of course Vane Permissions, but these optional hooks might need to be updated and people might ask for more permission plugins to be added (which of course might not be an actual issue). I haven't tried any sort of workaround to get permissions without hooking into these plugins, so if you have any ideas let me know.
### Custom Message & Translations on Resource Pack Confirmation Screen
We can add a message or let admins customize what message goes on the resource pack confirmation screen. I would look like this:
<img width="1404" height="444" alt="image" src="https://github.com/user-attachments/assets/1c5e59bc-d925-429c-8996-af2409f5aa91" />
Though I am unfamiliar with how Vane's language and translation system and also config file system works. Obviously the text above the custom message adhere's the Minecraft's language selection and it would be nice if we could query Vane to send the correct message based on the player's language settings.

Does Vane auto add config settings programmatically and automatically stick them onto existing users configs? Does the same thing work for lang messages? Any insight here would be great. 🙏 

---

Let me know your thoughts and we can go from there, thanks!

FaultyFunctions committed:
Removed ProtocolLib + Resource pack load during configuration phase initial setup

oddlama reviewed (commented):
> Please let me know if there is something wrong with my code.

Not at all, looks pretty good to me!

> An alternative might be to interface directly with popular permission plugins like LuckPerms and of course Vane Permissions, but these optional hooks might need to be updated and people might ask for more permission plugins to be added (which of course might not be an actual issue). I haven't tried any sort of workaround to get permissions without hooking into these plugins, so if you have any ideas let me know.

I think the bypass permission has mostly been requested and used by players who have used a bedrock client. If I recall correctly, it didn't support sending the usual resource pack status messages, so the server could never see that they accepted the pack. Unfortunately I don't know much about bedrock these days, so I'm not sure what is the status quo there.

In any case, vane does not officially support bedrock, and since this new way of configuring resource packs is definitely what Mojang intends to support in the future, I would not want to block this for that reason.

I'm not sure what Geyser (the bedrock<->java compat plugin) does with these new resource pack reconfiguration events. I think you still have to prepare a bedrock specific resource pack and put it somewhere, so maybe these events do not even get called by bedrock players. Ideally it would just work the way it is and we can just remove the old permission.

> We can add a message or let admins customize what message goes on the resource pack confirmation screen.

That would be awesome!

> Though I am unfamiliar with how Vane's language and translation system and also config file system works. Obviously the text above the custom message adhere's the Minecraft's language selection and it would be nice if we could query Vane to send the correct message based on the player's language settings.

I think this won't be possible in this specific case, because the message must be displayed before the resource pack is sent/loaded. So we don't have any customizable translated messages available on the client. But it would still be possible to use the language system of vane and render the string on the server. That would mean that all players will see the same message, but at least the server owner could select one language globally and the message would be translated.

> Does Vane auto add config settings programmatically and automatically stick them onto existing users configs? Does the same thing work for lang messages? Any insight here would be great. 🙏

Yep, I made the `@Config<whatever>` and `@Lang` annotations exactly for that purpose! New config options will just appear in existing configurations without any action by users. Only renaming or moving options around is not properly supported.

The lang system is unfortunately a bit worse than the configuration system. It follows the same principle but honestly I think I made some big design mistakes there xD. The language files are always copied into a directory `plugins/vane-<whatever>/lang`, and editing them is not as easy as editing configuration files since it requires a version tag update and a regeneration of the resource pack.

It would be far better to not have the language files in the plugins directory at all and just keep them in memory. If the new resource pack configuration phase allows streaming the resource pack directly from the server instead of serving an url, it could be generated on the fly on startup and woud always serve an up-to-date variant. Well, I guess all of that is for a potential v2 some time in the future...

FaultyFunctions committed:
Lang customization + cleanup

- Added custom prompts to lang file and resource pack prompt screen
- Removed vane.resourcepack.bypass
- Code/file cleanup
- Added PacketEvents to README acknowledgements
- Remove ProtocolLib dependency from build.gradle.kts

FaultyFunctions committed:
Swap to ConcurrentHashMap to ensure thread safety

FaultyFunctions committed:
Ensure old latches are released properly

oddlama reviewed (commented):
I just have one question about the removed exemption mechanism, other than that this looks great!

FaultyFunctions commented on vane-core/src/main/java/org/oddlama/vane/core/resourcepack/ResourcePackDistributor.java:
```diff
@@ -36,43 +41,31 @@ public class ResourcePackDistributor extends Listener<Core> {
     @ConfigBoolean(
         def = true,
-        desc = "Kick players if they deny to use the specified resource pack (if set). Individual players can be exempt from this rule by giving them the permission 'vane.core.resource_pack.bypass'."
+        desc = "Kick players if they deny to use the specified resource pack (if set)."
```
So the old system worked by logging the player in fully first and then kicking them if they didn't have bypass perms and they decline the RP. The new system works before they are fully logged in so there is no player object to check .hasPermission() on (I think only PlayerProfile is available). It's a similar issue to trying to check permissions on an offline player. It's "possible" to get it working, but we'd need to interface directly and explicitly with a permission plugin. So we'd have to support multiple permission plugins to cover all our bases and I feel like it's a lot of work and maintenance just for this one feature.

Unless I'm incorrect and there is an easier way to check permissions when we can't access the player object? Not sure, but let me know if you have more insight than me!

oddlama replied:
Not at all - makes total sense that we cannot check permissions on pre-logged-in players in a general way. If someone required that feature they can still disable the check for everyone, which I think is fair.

FaultyFunctions commented:
> Also if we don't need protocollib anymore, can you do a quick find-in-all-files in the codebase to see if there are any mentions left? I think it would be good to update the readme and other stuff that references it.

Sure! I'll do that when I get a chance today or tomorrow 👍

oddlama commented:
Totally forgot about this one 😅, any objections to merging this now?

FaultyFunctions commented:
No objections! Apologies I didn't have time to remove the protocollib mentions from the readme but should be functionally ready to merge!
